src/back/hlsl/writer.rs

```diff
@@ -1856,14 +1856,44 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                     }
                 };
 
-                let var_handle = self.fill_access_chain(module, pointer, func_ctx)?;
-                // working around the borrow checker in `self.write_expr`
-                let chain = mem::take(&mut self.temp_access_chain);
-                let var_name = &self.names[&NameKey::GlobalVariable(var_handle)];
+                let pointer_space = {
+                    let pointer_ty = match func_ctx.info[pointer].ty {
+                        proc::TypeResolution::Handle(handle) => &module.types[handle].inner,
+                        proc::TypeResolution::Value(ref inner) => inner,
+                    };
+                    match pointer_ty {
+                        &TypeInner::Pointer { space, .. } => space,
+                        &TypeInner::ValuePointer { space, .. } => space,
+                        ref other => {
+                            return Err(Error::Custom(format!(
+                                "unexpected resolved type for atomic pointer operand: {other:?}"
+                            )))
+                        }
+                    }
+                };
```

Comment:
```suggestion
                let pointer_space = func_ctx.info[pointer].ty
                    .inner_with(&module.types)
                    .pointer_space();
```

We have a few helper functions that we could use here.

Reply:
👀 Onice, implemented more-or-less as suggested with 57a9728b.